src/compiler/checker.ts

```diff
@@ -25183,11 +25199,17 @@ namespace ts {
                 return links.jsxNamespace;
             }
             if (!links || links.jsxNamespace !== false) {
-                const namespaceName = getJsxNamespace(location);
-                let resolvedNamespace = resolveName(location, namespaceName, SymbolFlags.Namespace, /*diagnosticMessage*/ undefined, namespaceName, /*isUse*/ false);
-                if (!resolvedNamespace || resolvedNamespace === unknownSymbol) {
+                let resolvedNamespace: Symbol | undefined;
+                let namespaceName = getLocalClassicJsxNamespace(location);
+                if (!namespaceName) {
                     resolvedNamespace = getJsxNamespaceContainerForImplicitImport(location);
                 }
+                if (!resolvedNamespace) {
+                    if (!namespaceName) {
+                        namespaceName = getConfigClassicJsxNamespace();
+                    }
+                    resolvedNamespace = resolveName(location, namespaceName, SymbolFlags.Namespace, /*diagnosticMessage*/ undefined, namespaceName, /*isUse*/ false);
+                }

```

Comment:
Rather than globally changing the priority of the implicit lookup, we should instead just prefer the implicit namespace if `jsx` is `react-jsx` or `react-jsxdev`, no?

```suggestion
                let resolvedNamespace: Symbol | undefined;
                if (compilerOptions.jsx === JsxEmit.ReactJSX || compilerOptions.jsx === JsxEmit.ReactJSXDev) {
                    resolvedNamespace = getJsxNamespaceContainerForImplicitImport(location);
                }
                else {
                    const namespaceName = getJsxNamespace(location);
                    resolvedNamespace = resolveName(location, namespaceName, SymbolFlags.Namespace, /*diagnosticMessage*/ undefined, namespaceName, /*isUse*/ false);
                }
```

Reply:
If `@jsxImportSource` should be prioritized over `@jsx` then - yes. I also believe that this is a sensible priority order.

I can revert my changes quickly and apply the proposed patch but would prefer doing that after establishing how a failing test (for master) should look like so it would be verified by this PR that this actually solves the problem but - as I've mentioned - I have no idea right now that's the difference between my attempt at providing that failing test and the presented repro case. Any tips regarding that?

Reply:
> If @jsxImportSource should be prioritized over @jsx then - yes. I also believe that this is a sensible priority order.

Yeah, I think it should be. New thing > old thing is usually a good priority order.

> Any tips regarding that?

The resolution behavior will change depending on what files are present "on disk" on the test. You may need to provide, eg, a `node_modules/react/index.d.ts` and/or a `node_modules/react/jsx-runtime.d.ts` (using minimal definitions for these, rather than the full `react16` lib reference is likely what you'd need), or equivalent for an import source.

Reply:
> The resolution behavior will change depending on what files are present "on disk" on the test. You may need to provide, eg, a node_modules/react/index.d.ts and/or a node_modules/react/jsx-runtime.d.ts (using minimal definitions for these, rather than the full react16 lib reference is likely what you'd need), or equivalent for an import source.

I'm not sure if I completely understand this but this is definitely a good starting point for further exploration. It's already midnight here so I will get to it tomorrow and let you know if I've managed to create such a test.

Reply:
@weswigham I've pushed out a good test case and a patch similar to what you have proposed. However, I've decided to call `getJsxNamespaceContainerForImplicitImport` unconditionally first because `@jsxImportSource` should be taken into account as well

Reply:
Hm, alright. It looks like `getJSXImplicitImportBase` already matches that behavior, and that's what controls emit, so that seems reasonable.

Reply:
Some tests of this would be good, though! So some tests where `// @jsx: react` but the file has a `jsxImportSource` pragma (indicating use of the new jsx transform).

Reply:
Oh, so you are one of those... Added a new test to cover for this 😉